src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs

```diff
@@ -33,24 +71,116 @@ public void InvokeClearCache_NoExceptions()
+                EventInfo ei1 = s_type.GetEvent(nameof(Event1));
+                ConstructorInfo ci1 = s_type.GetConstructor(Type.EmptyTypes);
+
+                AssertMembersAreNotNull(mi1, pi1, fi1, ei1, ci1);
+
+                clearCache(new[] { typeof(ReflectionCacheTests) });
+
+                MethodInfo mi2 = s_type.GetMethod(nameof(Method));
+                PropertyInfo pi2 = s_type.GetProperty(nameof(Property));
+                FieldInfo fi2 = s_type.GetField(nameof(Field1));
+                EventInfo ei2 = s_type.GetEvent(nameof(Event1));
+                ConstructorInfo ci2 = s_type.GetConstructor(Type.EmptyTypes);
+
+                AssertMembersAreNotNull(mi2, pi2, fi2, ei2, ci2);
+
+                // After the Cache cleared the references of the same members of the same type will be different
+                // But they should be evaluated as Equal so that there were no issue using the same member after hot reload
+                AssertMemberReferencesNotSameButEqual(mi1, pi1, fi1, ei1, ci1, mi2, pi2, fi2, ei2, ci2);
 
-            MethodInfo mi1 = typeof(ReflectionCacheTests).GetMethod(nameof(GetMethod_MultipleCalls_ClearCache_DifferentObjects));
-            Assert.NotNull(mi1);
-            Assert.Equal(nameof(GetMethod_MultipleCalls_ClearCache_DifferentObjects), mi1.Name);
+                // And the HashCode of a member before and after hot reload should produce same result 
+                AssertHashCodesAreEqual(mi1.GetHashCode(), pi1.GetHashCode(), fi1.GetHashCode(), ei1.GetHashCode(), ci1.GetHashCode(),
+                    mi2.GetHashCode(), pi2.GetHashCode(), fi2.GetHashCode(), ei2.GetHashCode(), ci2.GetHashCode());
+            }, options);
+        }
 
-            clearCache(justSpecificType ? new[] { typeof(ReflectionCacheTests) } : null);
+        private static void AssertMembersAreNotNull(MethodInfo mi, PropertyInfo pi, FieldInfo fi, EventInfo ei, ConstructorInfo ci)
+        {
+            Assert.NotNull(mi);
+            Assert.NotNull(pi);
+            Assert.NotNull(fi);
+            Assert.NotNull(ei);
+            Assert.NotNull(ci);
+        }
 
-            MethodInfo mi2 = typeof(ReflectionCacheTests).GetMethod(nameof(GetMethod_MultipleCalls_ClearCache_DifferentObjects));
-            Assert.NotNull(mi2);
-            Assert.Equal(nameof(GetMethod_MultipleCalls_ClearCache_DifferentObjects), mi2.Name);
+        private static void AssertHashCodesAreEqual(int mi1Hash, int pi1Hash, int fi1Hash, int ei1Hash, int ci1Hash,
+            int mi2Hash, int pi2Hash, int fi2Hash, int ei2Hash, int ci2Hash)
+        {
+            Assert.Equal(mi1Hash, mi2Hash);
+            Assert.Equal(pi1Hash, pi2Hash);
+            Assert.Equal(fi1Hash, fi2Hash);
+            Assert.Equal(ei1Hash, ei2Hash);
+            Assert.Equal(ci1Hash, ci2Hash);
+        }
 
+        private static void AssertMemberReferencesNotSameButEqual(MethodInfo mi1, PropertyInfo pi1, FieldInfo fi1, EventInfo ei1, ConstructorInfo ci1,
```

Comment:
Nit: These methods with a lot of arguments (in particular this one with 10 arguments) do look nice. Would it be better to have a method like this:

```
void AssertNotSameSameButEqual(object o1, object o2)
{
     Assert.NotSame(o1, o2);
     Assert.Equal(o1, o2);
     Assert.Equal(o1.GetHashCode(), o2.GetHashCode());
}
```
and call it on the individual pairs?

Reply:
Sure